Add query types for `BlockCommit` and `Vote`

OnedgeLee opened the pull request:
### Needs
Query for fetching recent validator's consensus activities is needed.

### Changes
Queries that fetch some `Vote` information from `BlockCommit` has been implemented.

### Related issues
- https://github.com/planetarium/libplanet/issues/2621

longfin commented on Libplanet.Explorer.Tests/Queries/VoteQueryTest.cs:
```diff
@@ -0,0 +1,289 @@
+        }
+
+        [Fact]
+        public async void CountVotesQuery()
```
```suggestion
        public async Task CountVotesQuery()
```

longfin commented on Libplanet.Explorer/Queries/VoteQuery.cs:
```diff
@@ -0,0 +1,246 @@
+    public class VoteQuery<T> : ObjectGraphType
+        where T : IAction, new()
+    {
+        public VoteQuery(IBlockChainContext<T> chainContext)
```
```suggestion
        public VoteQuery(BlockChain<T> blockChain, IStore store)
```

In most cases, `IBlockChainContext<T>` isn't necessary.  😅

longfin commented on Libplanet.Explorer/Queries/VoteQuery.cs:
```diff
@@ -0,0 +1,246 @@
+            return result;
+        }
+
+        internal IEnumerable<Vote> ListValidatorVotes(
```
I think that it may be okay without this method. 🤔

longfin commented on Libplanet.Explorer/Queries/VoteQuery.cs:
```diff
@@ -0,0 +1,246 @@
+            return ListVotes(desc, offset, limit).Where(vote => vote.Validator.Equals(publicKey));
+        }
+
+        internal IEnumerable<VoteCount> CountVotes(
```
How about unifying `CountVotes()` and `CountValidatorVotes()` and replacing its parameters from `bool`, `long`, `long?` to `IEnumerable<Vote>`?

longfin commented on Libplanet.Explorer/Queries/VoteQuery.cs:
```diff
@@ -0,0 +1,246 @@
+            var result = new List<Vote>();
+            foreach (var index in indexList)
+            {
+                var votes = store.GetBlockCommit(index.value)?.Votes;
+                if (votes is { })
+                {
+                    result = result.Concat(store.GetBlockCommit(index.value).Votes).ToList();
```
```suggestion
                    result.AddRange(store.GetBlockCommit(index.value).Votes);
```

greymistcube replied:
```suggestion
                if (store.GetBlockCommit(index.value)?.Votes is { } votes)
                {
                    result.AddRange(votes);
```

We already know `votes` is not `null`. 😗

longfin commented on Libplanet.Explorer/Queries/ExplorerQuery.cs:
```diff
@@ -25,6 +25,7 @@ public ExplorerQuery(IBlockChainContext<T> chainContext)
             Field<BlockQuery<T>>("blockQuery", resolve: context => new { });
             Field<TransactionQuery<T>>("transactionQuery", resolve: context => new { });
             Field<StateQuery<T>>("stateQuery", resolve: context => new { });
+            Field<VoteQuery<T>>("previousVoteQuery", resolve: context => new { });
```
```suggestion
            Field<VoteQuery<T>>("voteQuery", resolve: context => new { });
```

As the type suggests, `voteQuery` seems fine. (or `blockVoteQuery`?)

longfin commented on Libplanet.Explorer.Tests/Queries/VoteQueryTest.cs:
```diff
@@ -0,0 +1,289 @@
+        [InlineData(1)]
+        [InlineData(2)]
+        [InlineData(3)]
+        public async void ValidatorVotesQuery(int validatorIdx)
```
```suggestion
        public async Task ValidatorVotesQuery(int validatorIdx)
```

longfin commented on Libplanet.Explorer.Tests/Queries/VoteQueryTest.cs:
```diff
@@ -0,0 +1,289 @@
+        }
+
+        [Fact]
+        public async void VotesQuery()
```
```suggestion
        public async Task VotesQuery()
```

longfin commented on Libplanet.Explorer.Tests/Queries/VoteQueryTest.cs:
```diff
@@ -0,0 +1,289 @@
+        [InlineData(1)]
+        [InlineData(2)]
+        [InlineData(3)]
+        public async void CountValidatorVotesQuery(int validatorIdx)
```
```suggestion
        public async Task CountValidatorVotesQuery(int validatorIdx)
```

greymistcube commented on Libplanet.Explorer/Queries/VoteQuery.cs:
```diff
@@ -0,0 +1,246 @@
+            var result = new List<Vote>();
+            foreach (var index in indexList)
+            {
+                var votes = store.GetBlockCommit(index.value)?.Votes;
```
`IStore` only has `BlockCommit`s for caching. I think `BlockChain<T>.GetBlockCommit()` should be used instead.

longfin commented on Libplanet.Explorer/GraphTypes/VoteType.cs:
```diff
@@ -0,0 +1,43 @@
+
+            Field<NonNullGraphType<LongGraphType>>(
+                "Height",
+                description: "Hight of the consensus voted.",
```
```suggestion
                description: "Height of the consensus voted.",
```

longfin reviewed (changes requested):
The build seems to fail, could you check?

OnedgeLee commented:
I've removed `VoteQuery`, since it's too specific to be a part of `Libplanet`. Instead, I've added `BlockCommitType` as a part of `BlockType`, make it possible to query `Vote`s on `BlockQuery`.

OnedgeLee commented:
> The build seems to fail, could you check?

Fixed
